readme.md
# MultiCalculator++
This is a command line calculator for Windows 10. I made it for my personal use 
for easy and fast accessible calculations during my own software projects and 
practises.


## Getting started
1. Open Windows PowerShell:
	* Press __Shift + Right Mouse Button__ in the folder and then __Open PowerShell Here__
	* Press __Windows + R__ and enter __powershell__
2. Go to the directory with the binary (Easy with the first method)
3. Enter __.\MultiCalculator++.exe command argument1 argument2__ or __.\MultiCalculator++ command argument1 argument2__


## List of all commands
### Listing of all fast commands:
- Addition:				-p n_1 n_2
- Add n days to a date:	-adtd day month year daystoadd

### General command format:
- Addition:				plus n_1> <n_2

### Command type differents:
You can run fast commands only one time per runtime, while you can run normal 
commands several times per runtime.


## Program command structure
MultiCalculator++ Command Argument1 Argument2 Argument3 Argument4

argv[0]: MultiCalculator++
argv[1]: Command
argv[2]: Argument1
argv[3]: Argument2
argv[4]: Argument3
argv[5]: Argument4


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## v0.5.3
- Fixed wrong comments in readme.md

## v0.5.2
- Chhanged in the 'Add to date' function the type of the 'year' variable from 
  int to unsigned short.

## v0.5.1
- Fixed some typos

## v0.5.0
- Added check function around the program.
- Fixed some bugs in connection with it.
- 'Add' function will work now with '-p' and 'plus' command.
- 'Add to date' function will work now with '-adtd' command.

## v0.4.0
- Added MIT license to the project.
- Added readme.md
- Added authors.md
- Removed doc.txt

## v0.3.0
- Reworked program structures, documentations and functions.

## v0.2.0
- Added function to calculate _n_ days to a date: -addDTD day month year daysToAdd.

## v0.1.0
- Added basic addition function: -add number1 number2.


src/main.cpp
//#include <string>
#include "multiCal++.h"

int main(int argc, char *argv[]){
	const short maxArguments = 6; // Needs to be adjusted, if more than 4 arguments needed!

	if(program_Checks() == true)
	{
		bool programLoop = true;

		if(argc == 4 && string(argv[1]) == "-p")
		{
			float sum = add(stof(argv[2]), stof(argv[3]));
			cout	<< "The addition of "
					<< stof(argv[2])
					<< " and "
					<< stof(argv[3])
					<< " is "
					<< sum
					<< "\n"
					<< endl;
			programLoop = false;
		}
		if(argc == 6 && string(argv[1]) == "-adtd")
		{
			add_Days_To_Date(stous(argv[2]), stous(argv[3]), stoi(argv[4]), stoi(argv[5]));
		}
		if(argc > maxArguments)
		{
			cout	<< "Too many arguments! Currently allowed: "
					<< maxArguments
					<< endl;
		}
		if(argc > 2 && string(argv[1]) != "exit")
		{
			if(argc == 4 && string(argv[1]) == "plus")
			{
				while(programLoop == true)
				{
					float number1 = 0;
					float number2 = 0;
					cout << "Enter 2 numbers:\n" << endl;
					cin >> number1;
					cin >> number2;
					float sum = add(number1, number2);
					cout << endl;
					cout	<< "The addition of "
							<< number1
							<< " and "
							<< number2
							<< " is "
							<< sum
							<< "\n"
							<< endl;
					programLoop = continue_Program_Loop();
				}
			}
		}

		else
		{
			if(argc == 1)
			{
				cout	<< "No arguments passed. ";
				cout 	<< "Enter 'help' or 'usage' for more information.\n"
						<< endl;
			}
			if(argc == 2 && (string(argv[1]) == "help" || string(argv[1]) == "Help" || string(argv[1]) == "usage" || string(argv[1]) == "Usage"))
			{
				cout	<< "usage:\n"
						<< "-add <number1> <number2>\n"
						<< "-adtd <day> <month> <year> <daystoadd>"
						<< endl;
			}
		}
	}
	else cout << "Error: Program checks did not pass or failed unintentionally." << endl;

	return 0;
}


src/multiCal.cpp
#include "multiCal.h" // Link to header file

// Run in Windows 10:	.\multiCal.exe argvName variable1 variable2
//						.\multiCal argvName variable1 variable2
int main(int argc, char *argv[]){
	const short maxArguments = 6; // Needs to be adjusted, if more arguments can be used!


	// Command line scenarios
	if(argc == 1){ // argc=1 comes from the programm name
		// usage texts for the functions:
		cout	<< "No arguments passed. Please enter arguments. \n"
				<< endl;
		cout	<< "usage: \n"
				<< usage_basicAddition
				<< endl
				<< usage_addDaysToDate
				<< endl;
	}
	if(argc >= 2 && argc <= maxArguments){ // argv[0]=programm name, argv[1]=1. argument, argv[2]=2. arguments, etc.
		// basicAddition function:
		if(argv[1] == argvName_basicAddition){ // Compare strings
			basicAddition(stof(argv[2]), stof(argv[3])); // Use converted arguments with function; converted strings to floats
		}
		// addDaysToDate function:
		if(argv[1] == argvName_addDaysToDate){
			addDaysToDate(stous(argv[2]), stous(argv[3]), stoi(argv[4]), stoi(argv[5]));
		}
	}
	if(argc > maxArguments){
		cout	<< "Too many arguments! Currently allowed: "
				<< maxArguments
				<< endl;
	}


	return 0; // Statement for a successfull termination
}


src/multiCal++.cpp
#include "multiCal++.h"


// Calculation functions:
float add(float number1, float number2){
	return number1 + number2;
}


void add_Days_To_Date(unsigned short day, unsigned short month, unsigned short year, int daysToAdd)
{
	if(	((month == 1 || month == 3 || month == 5 || month == 7 || month == 8 || month == 10 || month == 12) && day > 0 && day < 32)
		|| ((month == 4 || month == 6 || month == 9 || month == 11) && day > 0 && day < 31)
		|| (month == 2 && day > 0 && day < 29 && year % 4 != 0 && year % 400 != 0)
		|| (month == 2 && day > 0 && day < 30 && (year % 4 == 0 || year % 400 == 0))
		) // Input handling
	{
		unsigned short storeDay = day;
		unsigned short storeMonth = month;
		unsigned short storeYear = year;
		int storeDaysToAdd = daysToAdd;


		if(daysToAdd >= 0) // If daysToAdd is positive or null...
		{
			for(int i = daysToAdd; i != 0; i--) // ...then compute days
			{
				if(month == 4 || month == 6 || month == 9 || month == 11)  // Handle 30 days months
				{
					if(day == 30)
					{
						day = 1;
						month++;
					}
					else day++;
				}
				else if(month == 2) // Handle februrary (month=2) and leap years
				{
					if(day == 28 && year % 4 != 0 && year % 400 != 0)
					{
						day = 1;
						month++;
					}
					else if(day == 29 && (year % 4 == 0 || year % 100 != 0 || year % 400 == 0))
					{
						day = 1;
						month++;
					}
					else day++;
				}
				else // Handle 31 days months: 1, 3, 5, 7, 8, 10, 12
				{
					if(day == 31)
					{
						if(month == 12) // December case
						{
							day = 1;
							month = 1;
							year++;
						}
						else // All other cases
						{
							day = 1;
							month++;
						}
					}
					else day++;
				}
			}
			cout	<< "\n"
					<< "Calculating "
					<< storeDaysToAdd
					<< " days to "
					<< storeDay
					<< "."
					<< storeMonth
					<< "."
					<< storeYear
					<< " results in: "
					<< day
					<< "."
					<< month
					<< "."
					<< year
					<< "\n"
					<< endl;
		}
		else // If daysToAdd is negative (daysToAdd < 0)
		{
			for(int i = daysToAdd; i != 0; i++) // ...then compute days
			{
				if(month == 4 || month == 6 || month == 9 || month == 11) // Handle 30 days months
				{
					if(day == 1) // Checked: All these month 1 back are always 31 days
					{
						month--;
						day = 31;
					}
					else{day--;}
				}
				else if(month == 2) // Handle februrary (month=2) and leap years
				{
					if(day == 1) // Then go back to january
					{
						day = 31;
						month--;
					}
					else day--;
				}
				else // Handle 31 days months: 1, 3, 5, 7, 8, 10, 12
				{
					if(day == 1)
					{
						if(month == 1) // January case
						{
							day = 31;
							month = 12;
							year--;
						}
						else if(month == 3 && year % 4 != 0 && year % 400 != 0) // March case; no leap year
						{
							day = 28;
							month--;
						}
						else if(month == 3 && (year % 4 == 0 || year % 100 != 0 || year % 400 == 0))  // March case; leap year
						{
							day = 29;
							month--;
						}
						else if(month == 8) // August case
						{
							day = 31;
							month--;
						}
						else  // 30 month cases: may, july, august, october, december
						{
							day = 30;
							month--;
						}
					}
					else day--;
				}
			}
			cout	<< "\n"
					<< "Calculate "
					<< storeDaysToAdd
					<< " days to "
					<< storeDay
					<< "."
					<< storeMonth
					<< "."
					<< storeYear
					<< " results in: "
					<< day
					<< "."
					<< month
					<< "."
					<< year
					<< "\n"
					<< endl;
		}
	}
	else
	{
		cout	<< "Error: Day must be between 1-31 and month must be beetween 1-12 (Please respect the right amount of days per month)!"
				<< endl;
	}
}


// Support functions:
bool program_Checks()
{
	bool b = true; // true = everything is OK

	// add-Function-Checks:
	if(add(0, 0) != 0) b = false;
	if(add(-5, -5) != -10) b = false;
	if(add(5, 5) != 10) b = false;
	if(add(5.5, 5.5) != 11) b = false;
	if(add(-5.5, -5.5) != -11) b = false;

	return b;
}


bool continue_Program_Loop()
{
	bool b = true;
	string s = "";

	cout << "Do you want to continue calculating?\n" << endl;
	cin >> s;
	if(s == "yes" || s == "Yes" || s == "y" || s == "Y") b = true;
	if(s == "no" || s == "No" || s == "n" || s == "N") b = false;

	return b;
}


unsigned short stous(string str)
{
	unsigned int uint = stoul(str); // Converting string into unsigned integer
	unsigned short s = (short)uint; // Converting/Cast unsigned integer to unsigned short
	return s;
}


src/multiCal.h
#ifndef SRC_MULTICAL_H_
#define SRC_MULTICAL_H_


#include <iostream> // Needed for in- and outputs
#include <string> // Needed for strings
using namespace std; // Automatic use of std, e.g. std::cout


// Main programm functions:
void basicAddition(float number1, float number2){
	cout	<< "\n"
			<< "The addition of "
			<< number1
			<< " and "
			<< number2
			<< " is: "
			<< number1 + number2
			<< "\n"
			<< endl;
}
const string argvName_basicAddition = "-add";
const string usage_basicAddition = "  multiCal -add number1 number2 \n";


void addDaysToDate(unsigned short day, unsigned short month, int year, int daysToAdd){
	if(	((month == 1 || month == 3 || month == 5 || month == 7 || month == 8 || month == 10 || month == 12) && day > 0 && day < 32)
		|| ((month == 4 || month == 6 || month == 9 || month == 11) && day > 0 && day < 31)
		|| (month == 2 && day > 0 && day < 29 && year % 4 != 0 && year % 400 != 0)
		|| (month == 2 && day > 0 && day < 30 && (year % 4 == 0 || year % 400 == 0))
		){ // Input handling
		unsigned short storeDay = day;
		unsigned short storeMonth = month;
		int storeYear = year;
		int storeDaysToAdd = daysToAdd;


		if(daysToAdd >= 0){ // If daysToAdd is positive or null...
			for(int i = daysToAdd; i != 0; i--){ // ...then compute days
				if(month == 4 || month == 6 || month == 9 || month == 11){ // Handle 30 days months
					if(day == 30){day = 1; month++;}
					else{day++;}
				}
				else if(month == 2){ // Handle februrary (month=2) and leap years
					if(day == 28 && year % 4 != 0 && year % 400 != 0){day = 1; month++;}
					else if(day == 29 && (year % 4 == 0 || year % 100 != 0 || year % 400 == 0)){day = 1; month++;}
					else{day++;}
				}
				else{ // Handle 31 days months: 1, 3, 5, 7, 8, 10, 12
					if(day == 31){
						if(month == 12){day = 1; month = 1; year++;} // December case
						else{day = 1; month++;} // All other cases
					}
					else{day++;}
				}
			}
			cout	<< "\n"
					<< "Calculate "
					<< storeDaysToAdd
					<< " days to "
					<< storeDay
					<< "."
					<< storeMonth
					<< "."
					<< storeYear
					<< " results in: "
					<< day
					<< "."
					<< month
					<< "."
					<< year
					<< "\n"
					<< endl;
		}
		else{ // If daysToAdd is negative (daysToAdd < 0)
			for(int i = daysToAdd; i != 0; i++){ // ...then compute days
				if(month == 4 || month == 6 || month == 9 || month == 11){ // Handle 30 days months
					if(day == 1){month--; day = 31;} // Checked: All these month 1 back are always 31 days
					else{day--;}
				}
				else if(month == 2){ // Handle februrary (month=2) and leap years
					if(day == 1){day = 31; month--;} // Then go back to january
					else{day--;}
				}
				else{ // Handle 31 days months: 1, 3, 5, 7, 8, 10, 12
					if(day == 1){
						if(month == 1){day = 31; month = 12; year--;} // January case
						else if(month == 3 && year % 4 != 0 && year % 400 != 0){day = 28; month--;} // March case; no leap year
						else if(month == 3 && (year % 4 == 0 || year % 100 != 0 || year % 400 == 0)){day = 29; month--;} // March case; leap year
						else if(month == 8){day = 31; month--;} // August case
						else{day = 30; month--;} // 30 month cases: may, july, august, october, december
					}
					else{day--;}
				}
			}
			cout	<< "\n"
					<< "Calculate "
					<< storeDaysToAdd
					<< " days to "
					<< storeDay
					<< "."
					<< storeMonth
					<< "."
					<< storeYear
					<< " results in: "
					<< day
					<< "."
					<< month
					<< "."
					<< year
					<< "\n"
					<< endl;
		}
	}
	else{
		cout	<< "Error: Day must be between 1-31 and month must be beetween 1-12 (Please respect the right amount of days per month)!"
				<< endl;
	}
}
const string argvName_addDaysToDate = "-addDTD";
const string usage_addDaysToDate = "  multiCal -addDTD day month year +/-daysToAdd \n";


// Helper and support functions:
unsigned short stous(string str){ // string to unsigned short
	unsigned int uint = stoul(str); // Converting str into unsigned integer
	unsigned short s = (short)uint; // Converting/Cast unsigned integer to unsigned short
	return s;
}


#endif /* SRC_MULTICAL_H_ */
